src/mlpack/methods/logistic_regression/logistic_regression_function_impl.hpp

```diff
@@ -170,6 +171,47 @@ void LogisticRegressionFunction<MatType>::Gradient(
       * (responses[i] - sigmoid) + regularization;
 }
 
+//! Evaluate the gradient of the logistic regression objective function.
+template<typename MatType>
+void LogisticRegressionFunction<MatType>::Gradient(
+    const arma::mat& parameters,
+    arma::mat& gradient,
+    arma::mat& derivative)
+{
+  // Regularization term.
+  arma::mat regularization;
+  regularization = lambda * parameters.col(0).subvec(1, parameters.n_elem - 1);
+
+  const arma::rowvec sigmoids = (1 / (1 + arma::exp(-parameters(0, 0)
+      - parameters.col(0).subvec(1, parameters.n_elem - 1).t() * predictors)));
+
+  derivative = sigmoids % (1 - sigmoids);
+
+  gradient.set_size(parameters.n_elem);
+  gradient[0] = -arma::accu(responses - sigmoids);
+  gradient.col(0).subvec(1, parameters.n_elem - 1) = -predictors * (responses -
+      sigmoids).t() + regularization;
+}
+
+//! Evaluate the hessian of the logistic regression function.
+template<typename MatType>
+void LogisticRegressionFunction<MatType>::Hessian(
+    const arma::mat& parameters,
+    const arma::mat& gradient,
+    const arma::mat& derivative,
+    arma::mat& hessian) const
+{
+  hessian.set_size(gradient.n_elem);
+  hessian(0) = gradient(0) * parameters(0);
+
+  arma::mat hessianTemp = arma::mat(
+      hessian.colptr(0) + 1, 1, gradient.n_elem - 1, false, true);
+  hessianTemp = (lambda *
+      (gradient.col(0).subvec(1, parameters.n_elem - 1).t() *
+      predictors % derivative)) * predictors.t() +
+      gradient.col(0).subvec(1, parameters.n_elem - 1).t();
```

Comment:
I think this is computing (according to the notation in the extended paper) `((d^i)^T \nabla^2 f(w^k) d^i)`, is that correct?  If so I think we could simplify the required interface for functions here.  Since `d` is from the TRON algorithm only and not specific to linear regression, I think it is possible to do the calculation above in `ConjugateGradient()` and restrict this to only compute `\nabla^2 f(w^k)`, which could let us simplify the interface in one of two ways:

```
// set both gradient and hessian
void Hessian(const mat& iterate, mat& gradient, mat& hessian);
```

or

```
void Gradient(const mat& iterate, mat& gradient);
void Hessian(const mat& iterate, mat& hessian);
```

Of course this comment only applies if my initial statement is true.